### PR TITLE
Fix relative_eq for small values

### LINKED ISSUE #65: relative_eq() has incorrect behavior for small values
Consider this small example:

```rust
let diff: f32 = 1.0e-7;
let x: f32 = 1.0e-5;
let y = x - diff;
assert_ne!(x, y);
relative_eq!(x, y) // returns true, but should be false
```

Note that `x` and `y` are normal, full-precision floating-point values (not "subnormals"), above the `epsilon` threshold, and their relative error is (x - y)/x, which equals 0.01, far above the default threshold. However, they are marked as "relatively equal" in the current implementation.

The problem is in the logic of the provided implementation of `RelativeEq::relative_eq()` for `f32` and `f64`, namely these lines:

https://github.com/brendanzab/approx/blob/308176d3d882c75dd9b57d9a7711e879a76119a8/src/relative_eq.rs#L65-L70

While the precise definition of relative equality and the parameters `epsilon` and `max_relative` are never clearly specified in the documentation, based on the referenced blog posts, it seems that the intention is to use `epsilon` as a zero-closeness threshold, and `max_relative` as the relative error threshold.

The issue stems from not checking the two input numbers (`self` and `other` in this case) individually against `epsilon` for a zero-threshold, but instead testing their *absolute difference* against `epsilon`. As a result, relative equality works as expected for large numbers, but in practice defaults to absolute-difference equality for values less than 1.

### PR DESCRIPTION
**Pull request by kitt159**
Change is described [here](https://github.com/brendanzab/approx/issues/65#issuecomment-3867555552)

Fixes #65

### CODE REVIEW

**Comment by jonaspleyer** [comment]
I am so sorry. I must have missed this PR completely.

I am a bit concerned about all the tests that were rewritten in this PR. What is the reason behind that?

**Comment by kitt159** [comment]
> I am a bit concerned about all the tests that were rewritten in this PR. What is the reason behind that?

The behavior changed, which required some tests to be modified. I also find the current implementation of the tests cleaner - there is less duplicate code. Some tests were written in such a way that they didn't actually test anything, for example because they tested the equality of identical numbers (implementation using == would also pass the test).

File: approx/tests/relative_eq.rs [hunk 1]
```diff
@@ -104,15 +96,15 @@ mod test_f32 {
...
-        assert_relative_eq!(-1e-40f32, 0.0f32, epsilon = 1e-40f32);
+        assert_relative_eq!(0.0f32, 1e-20f32, epsilon = 1e-20f32);
+        assert_relative_eq!(1e-20f32, 0.0f32, epsilon = 1e-20f32);
+        assert_relative_eq!(0.0f32, -1e-20f32, epsilon = 1e-20f32);
+        assert_relative_eq!(-1e-20f32, 0.0f32, epsilon = 1e-20f32);
 
-        assert_relative_ne!(1e-40f32, 0.0f32, epsilon = 1e-41f32);
-        assert_relative_ne!(0.0f32, 1e-40f32, epsilon = 1e-41f32);
-        assert_relative_ne!(-1e-40f32, 0.0f32, epsilon = 1e-41f32);
-        assert_relative_ne!(0.0f32, -1e-40f32, epsilon = 1e-41f32);
```
Comment (jonaspleyer) [inline approx/tests/relative_eq.rs:115]: This looks like a precision regression to me.

File: approx/tests/relative_eq.rs [hunk 2]
```diff
@@ -42,52 +42,44 @@ mod test_f32 {
         assert_relative_ne!(1.0f32, 1.0f32);
     }
 
+    fn compare_near_numbers(number: f32) {
+        let rel_eq_number = number * 1.0000001f32;
```
Comment (jonaspleyer) [inline approx/tests/relative_eq.rs]: this multiplication has an internal error associated with it. Consider this little program:

```rust
fn main() {
    let a = 1E-20;
    let b = 1E-10;
    println!("{}", a * b);
    println!("{}", 1E-30);
}
```
Output
```
0.0000000000000000000000000000009999999999999999
0.000000000000000000000000000001
```
How do we ensure that this multiplicative error does not invalidate our test (either via false positive or false negative).

**Review by jonaspleyer — COMMENTED** [review]
I have left some comments which I would like to discuss.

**Comment by jonaspleyer** [comment]
> > I am a bit concerned about all the tests that were rewritten in this PR. What is the reason behind that?
> 
> The behavior changed, which required some tests to be modified. I also find the current implementation of the tests cleaner - there is less duplicate code. Some tests were written in such a way that they didn't actually test anything, for example because they tested the equality of identical numbers (implementation using == would also pass the test).

I think that there is nothing wrong with testing cases which would be true via `==`. This is of course a sanity check which should be satisfied trivially but nothing wrong with having tests for that.

Your implementation of the tests is different since it involves a multiplication to calculate the numbers that should be compared.

File: approx/tests/relative_eq.rs (hunk 1 above)
Comment (kitt159) [reply approx/tests/relative_eq.rs:115]: 1e-40 is denormalized number, it is not safe to compare denormalized numbers using relative comparison. Therefore they are marked equal by absolute comparison.

File: approx/tests/relative_eq.rs (hunk 2 above)
Comment (kitt159) [reply approx/tests/relative_eq.rs]: I dont understand.

1) How can you tell if the result is correct? We are talking about approximate comparisons, there is no strict border between correct and incorrect behavior. 

2) Multiplication is defined by IEEE754 to return always correctly rounded result, so with given numbers it is just different way how to "hardcode" a number. In this case the result seems to be not correct but that is because input numbers are rounded.

1e-20 is stored as 9.9999999999999994515327145421e-21
1e-10 is stored as 1.0000000000000000364321973155e-10
Their product is most accurately represented as 9.99999999999999908174112566984e-31, not by 1e-30 which is stored as 1.00000000000000008333642060759E-30.

**Comment by kitt159** [comment]
> I think that there is nothing wrong with testing cases which would be true via ==. This is of course a sanity check which should be satisfied trivially but nothing wrong with having tests for that.

I prefer to have nonredundant tests but I can keep most of the tests unchanged if you want.

**Comment by jonaspleyer** [comment]
I have taken another longer look at your fix and I think that I will accept it into the codebase. However I will ask you to revert the changes done to the tests. I simply feel uncomfortable to change the test suite without any need.

If you want, you can include issue https://github.com/brendanzab/approx/pull/102 as a standalone test.

Once the changes are included, I will create a new release candidate and notify some folds within the ecosystem to test the new version (most notable ndarray). As long as they do not find anything that interferes with their current implementations, I will wait a bit longer and ultimately release the new version.

If you do not find the time to make the changes from above, simply let me know. I will then do them myself. Not a big deal.

And thanks again for your continued effort in this!

**Comment by jonaspleyer** [comment]
Can you confirm that you agree to license the contributions of this PR under the MIT license? We have an ongoing thread which is nearly resolved to add the MIT license to this crate to enable better compatibility with the Rust ecosystem:
https://github.com/brendanzab/approx/issues/94

**Comment by jonaspleyer** [comment]
Another question that has bugged me is the following: your solution requires the additional implementation of the `default_relative_eq` function in the `RelativeEq` trait. I have two conflicting questions in mind about this:

1. Can we circumvent this to enable better compatibility?
2. Is it better to introduce this separate function so that the breaking API is made clear and in order to avoid users running into problems unknowingly.

**Comment by kitt159** [comment]
I have minimized the changes to the tests. (Nevertheless, I think the tests could benefit from some refactoring. There is a lot of duplication.)

> 1. Can we circumvent this to enable better compatibility?
> 2. Is it better to introduce this separate function so that the breaking API is made clear and in order to avoid users running into problems unknowingly.

All I can say is that `relative_eq` requires two different numbers to work correctly. I am not sure how to properly incorporate those two numbers into this very generic implementation. Plus, I have realized I am not very good at thinking about backward compatibility issues... However, if you have an idea on how to improve the code in this regard, feel free to do so.
